File: daima/14/CustomPicker-Obj/CustomPicker/AnimalChooserViewController.m
//
//  AnimalChooserViewController.m
//  CustomPicker
//


#import "AnimalChooserViewController.h"

#define kComponentCount 2
#define kAnimalComponent 0
#define kSoundComponent 1

@implementation AnimalChooserViewController

@synthesize animalNames;
@synthesize animalSounds;
@synthesize animalImages;
@synthesize delegate;

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
	return kComponentCount;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView 
numberOfRowsInComponent:(NSInteger)component {
	if (component==kAnimalComponent) {
		return [self.animalNames count];
	} else {
		return [self.animalSounds count];
	}
}


- (UIView *)pickerView:(UIPickerView *)pickerView viewForRow:(NSInteger)row 
		  forComponent:(NSInteger)component reusingView:(UIView *)view {
	if (component==kAnimalComponent) {
		return [self.animalImages objectAtIndex:row];
	} else {
		UILabel *soundLabel;
		soundLabel=[[UILabel alloc] initWithFrame:CGRectMake(0,0,100,32)];
	//	[soundLabel autorelease];
		soundLabel.backgroundColor=[UIColor clearColor];
		soundLabel.text=[self.animalSounds objectAtIndex:row];
		return soundLabel;
	}
}

- (CGFloat)pickerView:(UIPickerView *)pickerView 
rowHeightForComponent:(NSInteger)component {
	return 55.0;
}

- (CGFloat)pickerView:(UIPickerView *)pickerView widthForComponent:(NSInteger)component {
	if (component==kAnimalComponent) {
		return 75.0;
	} else {
		return 150.0;
	}
}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row 
	   inComponent:(NSInteger)component {
    
    ViewController *initialView;
    initialView=(ViewController *)self.delegate;
    
    if (component==kAnimalComponent) {
        int chosenSound=[pickerView selectedRowInComponent:kSoundComponent];
        [initialView displayAnimal:[self.animalNames objectAtIndex:row]
                         withSound:[self.animalSounds objectAtIndex:chosenSound]
                     fromComponent:@"动物图像"];
    } else {
        int chosenAnimal=[pickerView selectedRowInComponent:kAnimalComponent];
        [initialView displayAnimal:[self.animalNames objectAtIndex:chosenAnimal]
                         withSound:[self.animalSounds objectAtIndex:row]
                     fromComponent:@"声音"];
    }

}


- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)didReceiveMemoryWarning
{
    // Releases the view if it doesn't have a superview.
    [super didReceiveMemoryWarning];
    
    // Release any cached data, images, etc that aren't in use.
}

#pragma mark - View lifecycle

/*
// Implement loadView to create a view hierarchy programmatically, without using a nib.
- (void)loadView
{
}
*/

-(void)viewWillDisappear:(BOOL)animated {
    ((ViewController *)self.delegate).animalChooserVisible=NO;
}

-(void)viewDidAppear:(BOOL)animated {
    ViewController *initialView;
    initialView=(ViewController *)self.delegate;
    [initialView displayAnimal:[self.animalNames objectAtIndex:0]
                     withSound:[self.animalSounds objectAtIndex:0]
                 fromComponent:@"还没有..."];
}

// Implement viewDidLoad to do additional setup after loading the view, typically from a nib.
- (void)viewDidLoad
{    
    self.animalNames=[[NSArray alloc]initWithObjects:
                      @"Mouse",@"Goose",@"Cat",@"Dog",@"Snake",@"Bear",@"Pig",nil];
	self.animalSounds=[[NSArray alloc]initWithObjects:
                       @"Oink",@"Rawr",@"Ssss",@"Roof",@"Meow",@"Honk",@"Squeak",nil];
	self.animalImages=[[NSArray alloc]initWithObjects:
                       [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"mouse.png"]],
                       [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"goose.png"]],
                       [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"cat.png"]],
                       [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"dog.png"]],
                       [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"snake.png"]],
                       [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"bear.png"]],
                       [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"pig.png"]],
                       nil
                       ];
    
    [super viewDidLoad];
}


- (void)viewDidUnload
{
    [self setDelegate:nil];
    [self setAnimalNames:nil];
    [self setAnimalImages:nil];
    [self setAnimalSounds:nil];
    [super viewDidUnload];
    // Release any retained subviews of the main view.
    // e.g. self.myOutlet = nil;
}

- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation
{
    // Return YES for supported orientations
    return (interfaceOrientation == UIInterfaceOrientationPortrait);
}

- (IBAction)dismissAnimalChooser:(id)sender {
    
    [self dismissViewControllerAnimated:YES completion:nil];
}
@end

File: daima/14/CustomPicker-Swift/CustomPicker/AnimalChooserViewController.swift
//
//  AnimalChooserViewController.swift


import UIKit

class AnimalChooserViewController: UIViewController, UIPickerViewDataSource, UIPickerViewDelegate {
    
    let kComponentCount: Int = 2
    let kAnimalComponent: Int = 0
    let kSoundComponent: Int = 1
    
    var animalNames: [String] = []
    var animalSounds: [String] = []
    var animalImages: [UIImageView] = []

    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        
        animalNames=["Mouse","Goose","Cat","Dog","Snake","Bear","Pig"]
        animalSounds=["Oink","Rawr","Ssss","Roof","Meow","Honk","Squeak"]
        animalImages=[
            UIImageView(image: UIImage(named: "mouse.png")),
            UIImageView(image: UIImage(named: "goose.png")),
            UIImageView(image: UIImage(named: "cat.png")),
            UIImageView(image: UIImage(named: "dog.png")),
            UIImageView(image: UIImage(named: "snake.png")),
            UIImageView(image: UIImage(named: "bear.png")),
            UIImageView(image: UIImage(named: "pig.png")),
        ]
        
        preferredContentSize = CGSize(width: 340,height: 380)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        let initialView: ViewController =
        presentingViewController as! ViewController
        initialView.displayAnimal(animalNames[0],
            withSound: animalSounds[0], fromComponent: "nothing yet...")
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        if (popoverPresentationController?.arrowDirection
            != UIPopoverArrowDirection.unknown) {
                
                // This view controller is running in a popover
                NSLog("I'm running in a Popover")
        }
    }
    
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return kComponentCount
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        if component==kAnimalComponent {
            return animalNames.count;
        } else {
            return animalSounds.count;
        }
    }
    
    func pickerView(_ pickerView: UIPickerView, rowHeightForComponent component: Int) -> CGFloat {
        return 55.0
    }
    
    func pickerView(_ pickerView: UIPickerView, widthForComponent component: Int) -> CGFloat {
        if component==kAnimalComponent {
            return 75.0;
        } else {
            return 150.0;
        }
    }
    
    func pickerView(_ pickerView: UIPickerView, viewForRow row: Int, forComponent component: Int, reusing view: UIView?) -> UIView {
        
        if component==kAnimalComponent {
            let chosenImageView: UIImageView = animalImages[row]
            let workaroundImageView: UIImageView = UIImageView(frame: chosenImageView.frame)
            workaroundImageView.backgroundColor = UIColor(patternImage: chosenImageView.image!)
            return workaroundImageView
        } else {
            let soundLabel: UILabel = UILabel(frame: CGRect(x: 0,y: 0,width: 100,height: 32))
            soundLabel.backgroundColor = UIColor.clear()
            soundLabel.text = animalSounds[row]
            return soundLabel
        }
    }
    
    func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
        
        let initialView: ViewController = presentingViewController as! ViewController
        
        if component==kAnimalComponent {
            let chosenSound: Int = pickerView.selectedRow(inComponent: kSoundComponent)
            initialView.displayAnimal(animalNames[row], withSound: animalSounds[chosenSound], fromComponent: "the Animal")
        } else {
            let chosenAnimal: Int = pickerView.selectedRow(inComponent: kAnimalComponent)
            initialView.displayAnimal(animalNames[chosenAnimal], withSound: animalSounds[row], fromComponent: "the Sound")
        }
    }
    
    @IBAction func dismissAnimalChooser(_ sender: AnyObject) {
        dismiss(animated: true , completion: nil)
    }



}
